espei/error_functions/chemical_potential_optimizer.py

```diff
@@ -0,0 +1,102 @@
+"""
+Author: Jonathan Siegel
+
+Contains classes and functions for optimizing at chemical potentials.
+"""
+
+import numpy as np
+from pycalphad.core.utils import point_sample
+from pycalphad.core.constants import MIN_SITE_FRACTION
+
+class PhaseConditions:
+  """
+  Contains data about a given phase at a given temperature and pressure.
+  Evaluates objectives, gradients and Hessians of the Gibbs Free energy.
+  Also contains information about the sublattices and site fractions.
+  This is passed to the optimization function as an input.
+  """
+  def __init__(self, models, phase, state_vars, obj_f, grad_f, hess_f, parameters):
+    self.state_vars = state_vars
+    self.obj_f = obj_f
+    self.grad_f = grad_f
+    self.hess_f = hess_f
+    self.phase = phase
+    self.parameters = np.array(list(parameters))
+    # Collect information about sublattices and which components they contain.
+    self.site_fraction_counts = []
+    self.site_ratios = {}
+    ind = 0
+    while True:
+      count = len([x for x in models[phase].site_fractions if x.sublattice_index == ind])
+      if count == 0:
+        break
+      self.site_fraction_counts.append(count)
+      ind += 1
```

Comment:
The site fraction counts are supposed to be the number of species in each sublattice, right?

If so, a better way to do this is using the `sublattice_dof` output from `generate_dof` function [from pycalphad](https://github.com/pycalphad/pycalphad/blob/6722a58410fb2c77e0fd59fa7cb1a8a8278e1841/pycalphad/core/utils.py#L160)

Reply:
Ok, I'll take a look at the sublattice_dof function.